### PR TITLE
Refactor image preprocessing iterators to subclass Sequence.

### PR DESCRIPTION
**Pull request by fchollet**
CC @Dref360

### CODE REVIEW

File: keras/preprocessing/image.py [hunk 1]
```diff
@@ -796,29 +821,19 @@ def __init__(self, x, y, image_data_generator,
...
-        # the indexing of each batch.
-        with self.lock:
-            index_array, current_index, current_batch_size = next(self.index_generator)
-        # The transformation of images is not under thread lock
-        # so it can be done in parallel
-        batch_x = np.zeros(tuple([current_batch_size] + list(self.x.shape)[1:]), dtype=K.floatx())
+    def _get_batches_of_transformed_samples(self, index_array):
+        batch_x = np.zeros(tuple([len(index_array)] + list(self.x.shape)[1:]),
+                           dtype=K.floatx())
         for i, j in enumerate(index_array):
```
Comment (Dref360) [inline keras/preprocessing/image.py]: Do not need the enumerate here. i is not used.

File: keras/preprocessing/image.py [hunk 2]
```diff
@@ -796,29 +821,19 @@ def __init__(self, x, y, image_data_generator,
...
             x = self.x[j]
             x = self.image_data_generator.random_transform(x.astype(K.floatx()))
             x = self.image_data_generator.standardize(x)
             batch_x[i] = x
         if self.save_to_dir:
-            for i in range(current_batch_size):
+            for i, j in enumerate(index_array):
```
Comment (Dref360) [inline keras/preprocessing/image.py]: enumerate is not needed here. i is not used

File: keras/preprocessing/image.py [hunk 3]
```diff
@@ -1037,10 +1057,10 @@ def next(self):
             batch_x[i] = x
         # optionally save augmented images to disk for debugging purposes
         if self.save_to_dir:
-            for i in range(current_batch_size):
+            for i, j in enumerate(index_array):
```
Comment (Dref360) [inline keras/preprocessing/image.py]: same

File: tests/keras/preprocessing/image_test.py [hunk 4]
```diff
@@ -53,10 +53,19 @@ def test_image_data_generator(self, tmpdir):
...
-                                       shuffle=True, save_to_dir=str(tmpdir)):
-                assert x.shape[1:] == images.shape[1:]
+                                       shuffle=True, save_to_dir=str(tmpdir),
+                                       batch_size=3):
+                assert x.shape == images[:3].shape
                 break
 
+            # Test `flow` behavior as Sequence
+            seq = generator.flow(images, np.arange(images.shape[0]),
+                                 shuffle=True, save_to_dir=str(tmpdir),
```
Comment (Dref360) [inline tests/keras/preprocessing/image_test.py]: Need a test that shuffle works and on_epoch_end works

File: keras/preprocessing/image.py (hunk 2 above)
Comment (fchollet) [reply keras/preprocessing/image.py]: i is used below to retrieve sample of index i in the batch, `batch_x[i]`

j is the index of the same sample in the original array, which we use in the filename

File: tests/keras/preprocessing/image_test.py (hunk 4 above)
Comment (fchollet) [reply tests/keras/preprocessing/image_test.py]: Added tests for shuffling and reshuffling upon `on_epoch_end`

**Pull request merged**